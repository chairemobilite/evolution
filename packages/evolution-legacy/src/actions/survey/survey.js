/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import bowser     from 'bowser';
//import 'whatwg-fetch';
//import 'promise-polyfill/src/polyfill';
import _get       from 'lodash/get';
import _set       from 'lodash/set';
import _cloneDeep from 'lodash/cloneDeep';
import isEqual    from 'lodash/isEqual';
import _unset     from 'lodash/unset';

const fetchRetry = require('@zeit/fetch-retry')(require('node-fetch'));
// TODO Default options for retry are as high as 15 seconds, during which the
// user get no feedback. Since update requests are queued in that time, are
// sequential and it is not possible to empty the queue, the requests will try
// to execute for n*15 seconds, during which time the user may have refreshed
// the page and will not see that the updated values have been udpated. With 4
// retries, the user will get feedback within 3 seconds and the queue will empty
// much faster, reducing the risk of invisible updates. The risk is still
// present if the server is quickly back online and the user is fast enough. See
// https://github.com/chairemobilite/transition/issues/1266
const fetch = async (url, opts) => {
    return await fetchRetry(url, Object.assign({ retry: { retries: 4 }, ...opts }));
}

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
// TODO This is not a react widget, so can't use the context here, or can we? TBD
import applicationConfiguration from 'evolution-frontend/lib/config/application.config';
import surveyHelper                                     from '../../helpers/survey/survey';
import * as surveyHelperNew                             from 'evolution-common/lib/utils/helpers';
import { incrementLoadingState, decrementLoadingState } from '../shared/loadingState.js';
import config                                           from 'chaire-lib-common/lib/config/shared/project.config';
import { updateSection as updateSectionTs, startUpdateInterview as startUpdateInterviewTs, updateInterview as updateInterviewTs } from 'evolution-frontend/lib/actions/Survey';

//export const setInterview = (interview) => ({
//  type: 'SET_INTERVIEW',
//  interviewLoaded: true,
//  interview
//});

// called whenever an update occurs in interview responses or when section is switched to
export const updateSection = updateSectionTs;
export const startUpdateInterview = startUpdateInterviewTs;
export const updateInterview = updateInterviewTs;

/**
 * Fetch an interview from server and set it for edition in validation mode.
 * 
 * @param {*} interviewUuid The uuid of the interview to open
 * @param {*} callback 
 * @returns 
 */
export const startSetSurveyValidateInterview = (interviewUuid, callback = function() {}) => {
  return (dispatch, getState) => {
    return fetch(`/api/survey/validateInterview/${interviewUuid}`, {
      credentials: 'include'
    })
    .then((response) => {

      if (response.status === 200) {
        response.json().then((body) => {
          if (body.interview)
          {
            const interview = body.interview;
            dispatch(startUpdateSurveyValidateInterview('home', {}, null, interview, callback));

          }
        });
      }
    })
    .catch((err) => {
      surveyHelperNew.devLog('Error fetching interview to validate.', err);
    });
  };
};

/**
 * Fetch an interview from server and set it for display in a one page summary.
 * 
 * TODO Only the section ('home', 'validationOnePager') is different from 'startSetSurveyValidateInterview' Re-use
 * 
 * @param {*} interviewUuid The uuid of the interview to open
 * @param {*} callback 
 * @returns 
 */
export const startSetValidateInterview = (interviewUuid, callback = function() {}) => {
  return (dispatch, getState) => {
    return fetch(`/api/survey/validateInterview/${interviewUuid}`, {
      credentials: 'include'
    })
    .then((response) => {

      if (response.status === 200) {
        response.json().then((body) => {
          if (body.interview)
          {
            const interview = body.interview;
            dispatch(startUpdateValidateInterview('validationOnePager', {}, null, interview, callback));

          }
        });
      }
    })
    .catch((err) => {
      surveyHelperNew.devLog('Error fetching interview to validate.', err);
    });
  };
};

/**
 * Fetch an interview from server and re-initialize the validated_data to the
 * participant's responses, but keeping the validation comments.
 *
 * @param {*} interviewUuid The uuid of the interview to open
 * @param {*} callback 
 * @returns 
 */
export const startResetValidateInterview = (interviewUuid, callback = function() {}) => {
  return (dispatch, getState) => {
    return fetch(`/api/survey/validateInterview/${interviewUuid}?reset=true`, {
      credentials: 'include'
    })
    .then((response) => {

      if (response.status === 200) {
        response.json().then((body) => {
          if (body.interview)
          {
            const interview     = body.interview;
            dispatch(startUpdateValidateInterview('validationOnePager', {}, null, interview, callback));
          }
        });
      }
    })
    .catch((err) => {
      surveyHelperNew.devLog('Error fetching interview to reset.', err);
    });
  };
};

export const startSetInterview = (activeSection = null, surveyUuid = undefined, history = undefined, preFilledResponses = undefined) => {

  return (dispatch, getState) => {
    const browserTechData = bowser.getParser(window.navigator.userAgent).parse();
    return fetch(`/api/survey/activeInterview/${surveyUuid ? `${encodeURI(surveyUuid)}` : ''}`, {
      credentials: 'include'
    })
    .then((response) => {
      if (response.status === 200) {
        response.json().then((body) => {
          if (body.interview)
          {
            let interview = body.interview;
            if (!activeSection)
            {
              for (let sectionShortname in applicationConfiguration.sections)
              {
                if (config.isPartTwo === true)
                {
                  if (applicationConfiguration.sections[sectionShortname].isPartTwoFirstSection === true)
                  {
                    activeSection = sectionShortname;
                    break;
                  }
                }
                else
                {
                  if (applicationConfiguration.sections[sectionShortname].previousSection === null)
                  {
                    activeSection = sectionShortname;
                    break;
                  }
                }
              }
            }
            const valuesByPath = {
              'responses._activeSection': activeSection
            };
            if (preFilledResponses) {
                Object.keys(preFilledResponses).forEach((key) => {
                    valuesByPath[`responses.${key}`] = preFilledResponses[key];
                })
            }
            // update browser data if different:
            const existingBrowserUa = _get(interview, 'responses._browser._ua', null);
            const newBrowserUa      = browserTechData._ua
            if (existingBrowserUa !== newBrowserUa)
            {
              valuesByPath['responses._browser'] = browserTechData;
            }
            dispatch(startUpdateInterview(activeSection, valuesByPath, null, interview));
          }
          else
          {
            dispatch(startCreateInterview(preFilledResponses));
          }
        });
      } else {
        console.log(`Get active interview: wrong responses status: ${response.status}`);
        if (history && response.status === 401)  {
          history.push('/unauthorized');
        }
      }
    })
    .catch((err) => {
      surveyHelperNew.devLog('Error fetching interview.', err);
    });
  };
};

export const startCreateInterview = (preFilledResponses = undefined) => {
  const browserTechData = bowser.getParser(window.navigator.userAgent).parse();
  return (dispatch, getState) => {
    return fetch('/api/survey/createInterview', {
      credentials: 'include'
    })
    .then((response) => {
      if (response.status === 200) {
        response.json().then((body) => {
          if (body.interview)
          {
            let activeSection = null;
            if (applicationConfiguration.sections['registrationCompleted'])
            {
              activeSection = 'registrationCompleted';
            }
            else
            {
              for (let sectionShortname in applicationConfiguration.sections)
              {
                if (applicationConfiguration.sections[sectionShortname].previousSection === null)
                {
                  activeSection = sectionShortname;
                  break;
                }
              }
            }
            const responses = {
              'responses._activeSection': activeSection,
              'responses._browser'      : browserTechData
            };
            if (preFilledResponses) {
                Object.keys(preFilledResponses).forEach((key) => {
                    responses[`responses.${key}`] = preFilledResponses[key];
                })
            }
            dispatch(startUpdateInterview(activeSection, responses, null, body.interview));
          }
          else
          {
            // we need to do something if no interview is returned (error)
          }
        });
      }
    })
    .catch((err) => {
      surveyHelperNew.devLog('Error creating interview.', err);
    });
  };
};

export const startUpdateSurveyValidateInterview = function(sectionShortname, valuesByPath = null, unsetPaths = null, interview = null, callback) {
  return {
    queue: 'UPDATE_INTERVIEW',
    callback: async function(next, dispatch, getState) {
      //surveyHelperNew.devLog(`Update interview and section with values by path`, valuesByPath);
      try {
        if (interview === null)
        {
          interview = _cloneDeep(getState().survey.interview);
        }

        //interview.sectionLoaded = null;

        dispatch(incrementLoadingState());

        if (valuesByPath && Object.keys(valuesByPath).length > 0)
        {
          surveyHelperNew.devLog(`Update interview and section with values by path`, JSON.parse(JSON.stringify(valuesByPath)));
        }

        //const oldInterview = _cloneDeep(interview);
        //const previousSection = surveyHelperNew.getResponse(interview, '_activeSection', null);

        const affectedPaths = {};

        if (Array.isArray(unsetPaths)) // unsetPaths if array (each path in array has to be deleted)
        {
          for (let i = 0, count = unsetPaths.length; i < count; i++)
          {
            const path          = unsetPaths[i];
            affectedPaths[path] = true;
            _unset(interview, path);
          }
        }
        else
        {
          unsetPaths = [];
        }

        // update language if needed:
        //const oldLanguage    = _get(interview, 'responses._language', null);
        //const actualLanguage = null;//i18n.language;
        //if (oldLanguage !== actualLanguage)
        //{
        //   valuesByPath['responses._language'] = actualLanguage;
        //}

        if (valuesByPath)
        {
          if (valuesByPath['_all'] === true)
          {
            affectedPaths['_all'] = true;
          }
          for (const path in valuesByPath)
          {
            affectedPaths[path] = true;
            _set(interview, path, valuesByPath[path]);
          }
        }

        sectionShortname = surveyHelperNew.getResponse(interview, '_activeSection', sectionShortname);
        //if (sectionShortname !== previousSection) // need to update all widgets if new section
        //{
        //  affectedPaths['_all'] = true;
        //}
        const updatedInterviewAndValuesByPath = updateSection(sectionShortname, interview, affectedPaths, valuesByPath);
        interview    = updatedInterviewAndValuesByPath[0];
        valuesByPath = updatedInterviewAndValuesByPath[1];

        if (!interview.sectionLoaded || interview.sectionLoaded !== sectionShortname)
        {
          valuesByPath['sectionLoaded'] = sectionShortname;
          interview.sectionLoaded = sectionShortname;
        }

        // convert undefined values to unset (delete) because stringify will remove undefined values:
        for (const path in valuesByPath)
        {
          if (valuesByPath[path] === undefined)
          {
            unsetPaths.push(path);
          }
        }

        if (isEqual(valuesByPath, {'_all': true}) && _isBlank(unsetPaths))
        {
          dispatch(updateInterview(_cloneDeep(interview)));
          dispatch(decrementLoadingState());
          if (typeof callback === 'function')
          {
            callback(interview);
          }
          return null;
        }

        //const differences = surveyHelper.differences(interview.responses, oldInterview.responses);
        const response = await fetch(`/api/survey/updateValidateInterview/${interview.uuid}`, {
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
          },
          credentials: 'include',
          method: "POST",
          body: JSON.stringify({
            id          : interview.id,
            participant_id     : interview.participant_id,
            valuesByPath: valuesByPath,
            unsetPaths  : unsetPaths,
            //responses  : interview.responses,
            //validations: interview.validations
          })
        });
        if (response.status === 200) {
          const body = await response.json();
          if (body.status == 'success' && body.interviewId == interview.uuid)
          {
            //surveyHelperNew.devLog('Interview saved to db');
            //setTimeout(function() {
              dispatch(updateInterview(_cloneDeep(interview)));
              if (typeof callback === 'function')
              {
                callback(interview);
              }
            //}, 500, 'That was really slow!');
          }
          else
          {
            // we need to do something if no interview is returned (error)
          }
        }
        // Loading state needs to be decremented, no matter the return value, otherwise the page won't get updated
        dispatch(decrementLoadingState());
      } catch(error) {
        console.log('Error updating interview', error);
        // Loading state needs to be decremented, no matter the return value, otherwise the page won't get updated
        // TODO Put in the finally block if we are sure there are no side effect in the code path that returns before the fetch
        dispatch(decrementLoadingState());
      } finally {
        next();
      }
    }
  };
};

export const startUpdateValidateInterview = function(sectionShortname, valuesByPath = null, unsetPaths = null, interview = null, callback) {
  return {
    queue: 'UPDATE_INTERVIEW',
    callback: async function(next, dispatch, getState) {
      try {
        if (interview === null)
        {
          interview = _cloneDeep(getState().survey.interview);
        }

        dispatch(incrementLoadingState());

        if (valuesByPath && Object.keys(valuesByPath).length > 0)
        {
          surveyHelperNew.devLog(`Update interview and section with values by path`, JSON.parse(JSON.stringify(valuesByPath)));
        }

        const affectedPaths = {};

        if (Array.isArray(unsetPaths)) // unsetPaths if array (each path in array has to be deleted)
        {
          for (let i = 0, count = unsetPaths.length; i < count; i++)
          {
            const path          = unsetPaths[i];
            affectedPaths[path] = true;
            _unset(interview, path);
          }
        }
        else
        {
          unsetPaths = [];
        }

        if (valuesByPath)
        {
          if (valuesByPath['_all'] === true)
          {
            affectedPaths['_all'] = true;
          }
          for (const path in valuesByPath)
          {
            affectedPaths[path] = true;
            _set(interview, path, valuesByPath[path]);
          }
        }

        sectionShortname = 'validationOnePager';
        const updatedInterviewAndValuesByPath = updateSection(sectionShortname, interview, affectedPaths, valuesByPath);
        interview    = updatedInterviewAndValuesByPath[0];
        valuesByPath = updatedInterviewAndValuesByPath[1];

        if (!interview.sectionLoaded || interview.sectionLoaded !== sectionShortname)
        {
          interview.sectionLoaded = sectionShortname;
        }

        // convert undefined values to unset (delete) because stringify will remove undefined values:
        for (const path in valuesByPath)
        {
          if (valuesByPath[path] === undefined)
          {
            unsetPaths.push(path);
          }
        }

        if (isEqual(valuesByPath, {'_all': true}) && _isBlank(unsetPaths))
        {
          dispatch(updateInterview(_cloneDeep(interview)));
          dispatch(decrementLoadingState());
          if (typeof callback === 'function')
          {
            callback(interview);
          }
          return null;
        }
        const response = await fetch(`/api/survey/updateValidateInterview/${interview.uuid}`, {
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
          },
          credentials: 'include',
          method: "POST",
          body: JSON.stringify({
            id          : interview.id,
            participant_id: interview.participant_id,
            valuesByPath: valuesByPath,
            unsetPaths  : unsetPaths,
          })
        })

        if (response.status === 200) {
          const body = await response.json();
          if (body.status === 'success' && body.interviewId === interview.uuid)
          {
            //surveyHelperNew.devLog('Interview saved to db');
            //setTimeout(function() {
              dispatch(updateInterview(_cloneDeep(interview)));
              dispatch(decrementLoadingState());
              if (typeof callback === 'function')
              {
                callback(interview);
              }
            //}, 500, 'That was really slow!');
          }
          else
          {
            dispatch(decrementLoadingState());
            // we need to do something if no interview is returned (error)
          }
        }
      } catch(error) {
        console.log('Error updating interview', error);
      } finally {
        next();
      }
    }

  };
};

export const startAddGroupedObjects = (newObjectsCount, insertSequence, path, attributes = [], callback, returnOnly = false) => {
  surveyHelperNew.devLog(`Add ${newObjectsCount} grouped objects for path ${path} at sequence ${insertSequence}`);
  return (dispatch, getState) => {
    const interview           = _cloneDeep(getState().survey.interview); // needed because we cannot mutate state
    const changedValuesByPath = surveyHelper.addGroupedObjects(interview, newObjectsCount, insertSequence, path, attributes || []);
    if (returnOnly)
    {
      return changedValuesByPath;
    }
    else
    {
      dispatch(startUpdateInterview(null, changedValuesByPath, null, null, callback));
    }
  };
};

export const startRemoveGroupedObjects = function(paths, callback, returnOnly = false) {
  surveyHelperNew.devLog(`Remove grouped objects at paths`, paths);
  return (dispatch, getState) => {
    const interview    = _cloneDeep(getState().survey.interview); // needed because we cannot mutate state
    let   unsetPaths   = [];
    let   valuesByPath = {};
    [valuesByPath, unsetPaths] = surveyHelper.removeGroupedObjects(paths, interview);
    if (returnOnly)
    {
      return [valuesByPath, unsetPaths];
    }
    else
    {
      dispatch(startUpdateInterview(null, valuesByPath, unsetPaths, null, callback));
    }
  };
};

export const startValidateAddGroupedObjects = (newObjectsCount, insertSequence, path, attributes = [], callback, returnOnly = false) => {
  surveyHelperNew.devLog(`Add ${newObjectsCount} grouped objects for path ${path} at sequence ${insertSequence}`);
  return (dispatch, getState) => {
    const interview           = _cloneDeep(getState().survey.interview); // needed because we cannot mutate state
    const changedValuesByPath = surveyHelper.addGroupedObjects(interview, newObjectsCount, insertSequence, path, attributes || []);
    if (returnOnly)
    {
      return changedValuesByPath;
    }
    else
    {
      dispatch(startUpdateValidateInterview(null, changedValuesByPath, null, null, callback));
    }
  };
};

export const startSurveyValidateRemoveGroupedObjects = function(paths, callback, returnOnly = false) {
  surveyHelperNew.devLog(`Remove grouped objects at paths`, paths);
  return (dispatch, getState) => {
    const interview    = _cloneDeep(getState().survey.interview); // needed because we cannot mutate state
    let   unsetPaths   = [];
    let   valuesByPath = {};
    [valuesByPath, unsetPaths] = surveyHelper.removeGroupedObjects(paths, interview);
    if (returnOnly)
    {
      return [valuesByPath, unsetPaths];
    }
    else
    {
      dispatch(startUpdateSurveyValidateInterview(null, valuesByPath, unsetPaths, null, callback));
    }
  };
};

export const startSurveyValidateAddGroupedObjects = (newObjectsCount, insertSequence, path, attributes = [], callback, returnOnly = false) => {
  surveyHelperNew.devLog(`Add ${newObjectsCount} grouped objects for path ${path} at sequence ${insertSequence}`);
  return (dispatch, getState) => {
    const interview           = _cloneDeep(getState().survey.interview); // needed because we cannot mutate state
    const changedValuesByPath = surveyHelper.addGroupedObjects(interview, newObjectsCount, insertSequence, path, attributes || []);
    if (returnOnly)
    {
      return changedValuesByPath;
    }
    else
    {
      dispatch(startUpdateSurveyValidateInterview(null, changedValuesByPath, null, null, callback));
    }
  };
};

export const startValidateRemoveGroupedObjects = function(paths, callback, returnOnly = false) {
  surveyHelperNew.devLog(`Remove grouped objects at paths`, paths);
  return (dispatch, getState) => {
    const interview    = _cloneDeep(getState().survey.interview); // needed because we cannot mutate state
    let   unsetPaths   = [];
    let   valuesByPath = {};
    [valuesByPath, unsetPaths] = surveyHelper.removeGroupedObjects(paths, interview);
    if (returnOnly)
    {
      return [valuesByPath, unsetPaths];
    }
    else
    {
      dispatch(startUpdateValidateInterview(null, valuesByPath, unsetPaths, null, callback));
    }
  };
};


