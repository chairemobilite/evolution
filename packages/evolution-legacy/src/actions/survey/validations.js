/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

export const checkValidations = function(validations, value, customValue, interview, path, customPath) {
  if (typeof validations === 'function')
  {
    try {
        const validationsGroup = validations(value, customValue, interview, path, customPath);
        for (let i = 0; i < validationsGroup.length; i++)
        {
          if (validationsGroup[i].validation === true)
          {
            return [false, validationsGroup[i].errorMessage];
          }
        }
      } catch (error) {
          // If there is an error during validations, just ignore the error and consider the value as valid to make sure we can at least get the next answers without blocking the questionnaire.
          // TODO: add a server-side log of the error and define a better way to deal with errors.
          console.log('validation error', error);
          return [true, null];
      }
  }
  return [true, null];
};

export const validateAllWidgets = function(interview) {
  return interview.allWidgetsValid;
  //let allWidgetsValid = true;
  //if (interview.widgets)
  //{
  //  for (let widgetPath in interview.widgets)
  //  {
  //    const widgetStatus = interview.widgets[widgetPath];
  //    if (widgetStatus.isVisible && !widgetStatus.isValid)
  //    {
  //      surveyHelper.devLog('invalid path: ', widgetPath);
  //      allWidgetsValid = false;
  //    }
  //  }
  //}
  //if (interview.groups)
  //{
  //  for (let groupShortname in interview.groups)
  //  {
  //    for (let groupedObjectId in interview.groups[groupShortname])
  //    {
  //      for (let widgetShortname in interview.groups[groupShortname][groupedObjectId])
  //      {
  //        const widgetStatus = interview.groups[groupShortname][groupedObjectId][widgetShortname];
  //        if (widgetStatus.isVisible && !widgetStatus.isValid)
  //        {
  //          surveyHelper.devLog('invalid widget in groupedObjectId ' + groupShortname + '.' + groupedObjectId + ':', widgetShortname);
  //          allWidgetsValid = false;
  //        }
  //      }
  //    }
  //  }
  //}
  //return allWidgetsValid;
};