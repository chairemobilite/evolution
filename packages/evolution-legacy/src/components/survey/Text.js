/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React               from 'react';
import { withTranslation } from 'react-i18next';
import Markdown            from 'react-markdown';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import * as surveyHelper from 'evolution-frontend/lib/utils/helpers';

export const Text = ({widgetConfig, widgetStatus, i18n, interview, path, user}) => {

  if(!widgetStatus.isVisible)
  {
    return null;
  }

  const content = surveyHelper.parseString(widgetConfig.text[i18n.language] || widgetConfig.text, interview, path, user);
  if (_isBlank(content))
  {
    return null;
  }

  return (<div className="survey-section__text">
    { widgetConfig.containsHtml && <div 
      dangerouslySetInnerHTML={{__html: content}}
    />}
    {!widgetConfig.containsHtml && <Markdown 
      className = {(widgetConfig.align || 'left') + ' ' + (widgetConfig.classes ? widgetConfig.classes : '')}
      source    = {content} 
    />}
  </div>);
};

export default withTranslation()(Text)