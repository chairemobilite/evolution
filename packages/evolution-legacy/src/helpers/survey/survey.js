/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { v4 as uuidV4 } from 'uuid';
import _get from 'lodash/get';
import sortBy from 'lodash/sortBy';

import sharedHelper from '../shared/shared';
import * as Helpers from 'evolution-common/lib/utils/helpers';
import * as LE from 'chaire-lib-common/lib/utils/LodashExtensions';
import * as DateTimeUtils from 'chaire-lib-common/lib/utils/DateTimeUtils';

export default {
  
  /**
   * @deprecated Use the LodashExtension function directly
   */
  isBlank                      : LE._isBlank,
  /**
   * @deprecated Use the LodashExtension function directly
   */
  chunkify                     : LE._chunkify,
  /**
   * @deprecated This does not seem used
   */
  differences                  : sharedHelper.differences,
  /**
   * @deprecated Use the evolution-common/lib/utils/helpers function directly
   */
  getPath                      : Helpers.getPath,
  /**
   * @deprecated Use the evolution-common/lib/utils/helpers function directly
   */
  devLog                       : Helpers.devLog,
  /**
   * @deprecated Use the evolution-common/lib/utils/helpers function directly
   */
  parseValue                   : Helpers.parseValue,
  /**
   * @deprecated Use the DateTimeUtils function directly
   */
  secondsSinceMidnightToTimeStr: DateTimeUtils.secondsSinceMidnightToTimeStr,
  /**
   * @deprecated Use the DateTimeUtils function directly
   */
  timeStrToSecondsSinceMidnight: DateTimeUtils.timeStrToSecondsSinceMidnight, 
  /**
   * @deprecated Use the evolution-common/lib/utils/helpers.getResponse function directly
   */
  get                          : Helpers.getResponse,
  /**
   * @deprecated Use the evolution-common/lib/utils/helpers.setResponse function directly
   */
  set                          : Helpers.setResponse,
  /**
   * @deprecated Use the evolution-common/lib/utils/helpers.getValidation function directly
   */
  getValidations: Helpers.getValidation,
  
  /**
   * @deprecated This does not seem used. DO NOT CALL, or open an issue to support it again
   */
  getRoot: function(interview, path, defaultValue = undefined) {
    return _get(interview, path, defaultValue);
  },

  /**
   * @deprecated Use the evolution-common/lib/utils/helpers.getValidation function directly
   */
  formatGeocodingQueryStringFromMultipleFields: Helpers.formatGeocodingQueryStringFromMultipleFields,

  // verify if value is a function. If it is, parse the value with the function, otherwise, return the string
  parseString: Helpers.parseString,

  parse: Helpers.parse,

  parseBoolean: Helpers.parseBoolean,

  parseInteger: Helpers.parseInteger,

  generateCurvedGooglePolyline: function(google, firstPoint /* google.maps.latLng */, lastPoint /* google.maps.latLng */, angle) {
    if(google.maps.geometry.spherical !== undefined)
    {
      var polylineCurvedPoints = [];
      var resolution           = 0.01;
      var distance             = google.maps.geometry.spherical.computeDistanceBetween(firstPoint, lastPoint);
      var heading              = google.maps.geometry.spherical.computeHeading(firstPoint, lastPoint);
      if(angle === undefined)
      {
        angle = 15.0;
      }
      var startingCurvedHeading = -angle + angle * (distance / 10000000);
      var nextPoint             = firstPoint;
      for(var point = 0; point < 0.500001; point += resolution)
      {
        polylineCurvedPoints.push(google.maps.geometry.spherical.computeOffset(nextPoint, distance * point * 2, heading + startingCurvedHeading * (2 * (0.5 - point))));
      }
      return polylineCurvedPoints;
    }
    return null;
  },

  interpolatePath: Helpers.interpolatePath,

  

  addGroupedObjects: function(interview, newObjectsCount, insertSequence, path, attributes = [])
  {
    const changedValuesByPath = {};
    const groupedObjects      = _get(interview.responses, path, {});
    const groupedObjectsArray = sortBy(Object.values(groupedObjects),['_sequence']);
    if (LE._isBlank(insertSequence) || insertSequence === -1)
    {
      insertSequence = groupedObjectsArray.length + 1;
    }
    else {
      // increment sequences of groupedObjects after the insertSequence:
      for(let seq = insertSequence, count = groupedObjectsArray.length; seq <= count; seq++)
      {
        const groupedObject = groupedObjectsArray[seq - 1];
        changedValuesByPath[`responses.${path}.${groupedObject._uuid}._sequence`] = seq + newObjectsCount;
      }
    }
    for (let i = 0; iÂ < newObjectsCount; i++)
    {
      const uniqueId            = uuidV4();
      const newSequence         = insertSequence + i;
      const newObjectAttributes = attributes[i] ? attributes[i] : {};
      changedValuesByPath[`responses.${path}.${uniqueId}`] = {'_sequence': newSequence, '_uuid': uniqueId, ...newObjectAttributes};
      changedValuesByPath[`responses.${path}.${uniqueId}`] = {'_sequence': newSequence, '_uuid': uniqueId, ...newObjectAttributes};
      changedValuesByPath[`validations.${path}.${uniqueId}`] = {};
    }
    return changedValuesByPath;
  },

  removeGroupedObjects: function(paths, interview)
  {
    // allow single path:
    if (!Array.isArray(paths))
    {
      paths = [paths];
    }

    if (paths.length === 0) {
        return [{}, []];
    }

    const unsetPaths   = [];
    const valuesByPath = {};
    let pathRemovedCount = 0;

    const groupedObjects        = Helpers.getResponse(interview, paths[0], {}, '../');
    const groupedObjectsArray   = sortBy(Object.values(groupedObjects),['_sequence']);

    for (let i = 0, count = groupedObjectsArray.length; i < count; i++) {
        const groupedObject = groupedObjectsArray[i];
        const groupedObjectPath = Helpers.getPath(paths[0], `../${groupedObject._uuid}`);
        if (paths.includes(groupedObjectPath)) {
            unsetPaths.push(`responses.${groupedObjectPath}`);
            unsetPaths.push(`validations.${groupedObjectPath}`);
            pathRemovedCount++;
        } else {
            if (pathRemovedCount > 0) {
                valuesByPath['responses.' + groupedObjectPath + '._sequence'] = groupedObject._sequence - pathRemovedCount;
            }
        }
    }
    return [valuesByPath, unsetPaths];
  },

  validateButtonAction: function(section, sections, saveCallback) {
    this.props.startUpdateInterview(section, { '_all': true }, null, null, function(interview) {
      if (interview.allWidgetsValid)
      {
        if (typeof saveCallback === 'function')
        {
          saveCallback();
        }
        else // go to next section
        {
          window.scrollTo(0, 0);
          this.props.startUpdateInterview(section, {
            'responses._activeSection': sections[section].nextSection
          });
        }
      }
      else
      {
        //this.props.startUpdateInterview(section, { '_all': true }, null, null, null);
      }
    }.bind(this));

    
  },

  validateButtonActionWithCompleteSection: function(section, sections, saveCallback) {
    this.props.startUpdateInterview(section, { '_all': true }, null, null, function(interview) {
      if (interview.allWidgetsValid)
      {
        if (typeof saveCallback === 'function')
        {
          saveCallback();
        }
        else // go to next section
        {
          window.scrollTo(0, 0);
          this.props.startUpdateInterview(section, {
            'responses._activeSection': sections[section].nextSection,
            [`responses._sections.${section}._isCompleted`]: true
          });
        }
      }
      else
      {
        //this.props.startUpdateInterview(section, { '_all': true }, null, null, null);
      }
    }.bind(this));

    
  },

  getWidgetFromPath(path)
  {
    // not yet implemented
    return null;
  },

  getWidgetChoiceFromValue(widget, value, interview)
  {
    let choices = widget.choices;
    if (typeof choices === 'function')
    {
      choices = choices(interview); // we should also get path for some questions, but ignoring for now...
    }
    if (!choices)
    {
      return null;
    }
    return choices.find(function(choice) {
      return choice.value === value;
    });
  }


};