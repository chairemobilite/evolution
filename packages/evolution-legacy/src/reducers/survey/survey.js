/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
export default (state = {}, action) => {
  switch (action.type) {
    case 'SET_INTERVIEW':
      return {
        interview:       action.interview,
        interviewLoaded: action.interviewLoaded,
      };
    case 'UPDATE_INTERVIEW':
      return {
        interview:       action.interview,
        interviewLoaded: action.interviewLoaded,
        errors:          action.errors,
        submitted:       action.submitted
      };
    default:
      return state;
  }
};