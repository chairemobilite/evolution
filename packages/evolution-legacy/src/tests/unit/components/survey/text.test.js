/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';

import {Text} from '../../../../components/survey/Text';
import {homeIntro} from '../../../fixtures/survey/widgets';

test('Should correctly render text widget', () =>{

    const i18n = {language: 'en'};

    const wrapper = TestRenderer.create(<Text i18n={i18n} widgetConfig={homeIntro} widgetStatus={{}} />);
    expect(wrapper).toMatchSnapshot();
});