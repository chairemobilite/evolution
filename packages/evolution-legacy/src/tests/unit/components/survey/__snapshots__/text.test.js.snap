// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render text widget 1`] = `null`;
